### PR TITLE
including two flap case

### PR DESCRIPTION
**Pull request by carme-hp**
With this modifications, the dealii-adapter would support the tutorial case with two perpendicular flaps. Besides FSI3 and PF, now we have the case PFleft and PFright, who only differ from PF by the x coordinates in point_bottom and point_tip.

### CODE REVIEW

**Review by MakisH — APPROVED** [review]
I think this is ready to review. It compiles and seems to be running with https://github.com/precice/tutorials/pull/110.

The code looks reasonable.

@DavidSCN what do you think? Do we need anything else?

**Review by MakisH — CHANGES_REQUESTED** [review]
Actually, we need an additional step: it looks like we pass the wrong vertices to `setMeshVertices` and each flap is still located in the middle.

@DavidSCN where can we modify the `support_points`? It looks like you call `DoFTools::map_dofs_to_support_points()` in `adapter.h`, but I am not sure what is going on under the hood. Maybe you have an idea?

**Comment by DavidSCN** [comment]
> @DavidSCN where can we modify the support_points? It looks like you call DoFTools::map_dofs_to_support_points() in adapter.h, but I am not sure what is going on under the hood. Maybe you have an idea?

I will have a look at the deal.II side later.

**Comment by DavidSCN** [comment]
> Actually, we need an additional step: it looks like we pass the wrong vertices to setMeshVertices and each flap is still located in the middle.

How did you come to this conclusion? Running both cases works just as expected for me. How can I reproduce the issue?

**Comment by MakisH** [comment]
> How did you come to this conclusion? Running both cases works just as expected for me. How can I reproduce the issue?

Indeed, here is the confusing part:
- deal.II writes the mesh in the correct locations
- the meshes exported by preCICE are in the wrong side

You can reproduce this by adding `<export:vtk directory="preCICE-output" />` in the solid1 / solid2 participants in the `precice-config.xml` and loading the surface meshes in ParaView (and visualizing them e.g. with glyphs).

**Comment by uekerman** [comment]
Should we unify the definition of `PF`, `PFleft` and `PFright` to avoid the code duplication? In the end only the location in x direction is different, right?

**Comment by DavidSCN** [comment]
> You can reproduce this by adding <export:vtk directory="preCICE-output" /> in the solid1 / solid2 participants in the precice-config.xml and loading the surface meshes in ParaView (and visualizing them e.g. with glyphs).

Done, looks for me reasonable. Also, selecting other files results in two separate beams.



> Should we unify the definition of `PF`, `PFleft` and `PFright` to avoid the code duplication? In the end only the location in x direction is different, right?

Yes. I think a more general 'mesh interface' without the duplication in each code is even a better approach and should be added in case this becomes larger.

**Comment by carme-hp** [comment]
That's weird, in my case if I visualize in paraview the same files you did in the screenshot the flaps are located in the same position. Can this be a paraview related issue? in my case I use 5.7.0

**Comment by DavidSCN** [comment]
If you have doubts about paraview you could directly look into the vtks. 
Running `head -n 20 Solid1_mesh-Fluid.init.vtk` prints the following:
```
# vtk DataFile Version 2.0

ASCII

DATASET UNSTRUCTURED_GRID

POINTS 157 float 

-1.0500000000000000e+00  0.0000000000000000e+00  0.0000000000000000e+00
-1.0500000000000000e+00  5.5555555555555552e-02  0.0000000000000000e+00
-1.0500000000000000e+00  9.5929535914450782e-03  0.0000000000000000e+00
-1.0500000000000000e+00  2.7777777777777776e-02  0.0000000000000000e+00
-1.0500000000000000e+00  4.5962601964110471e-02  0.0000000000000000e+00
-9.4999999999999996e-01  0.0000000000000000e+00  0.0000000000000000e+00
-9.4999999999999996e-01  5.5555555555555552e-02  0.0000000000000000e+00
-9.4999999999999996e-01  9.5929535914450782e-03  0.0000000000000000e+00
-9.4999999999999996e-01  2.7777777777777776e-02  0.0000000000000000e+00
-9.4999999999999996e-01  4.5962601964110471e-02  0.0000000000000000e+00
-1.0500000000000000e+00  1.1111111111111110e-01  0.0000000000000000e+00
-1.0500000000000000e+00  6.5148509147000627e-02  0.0000000000000000e+00
```
and `head -n 20 Solid2_mesh-Fluid.init.vtk` the following:
```
# vtk DataFile Version 2.0

ASCII

DATASET UNSTRUCTURED_GRID

POINTS 157 float 

9.4999999999999996e-01  0.0000000000000000e+00  0.0000000000000000e+00
9.4999999999999996e-01  5.5555555555555552e-02  0.0000000000000000e+00
9.4999999999999996e-01  9.5929535914450782e-03  0.0000000000000000e+00
9.4999999999999996e-01  2.7777777777777776e-02  0.0000000000000000e+00
9.4999999999999996e-01  4.5962601964110471e-02  0.0000000000000000e+00
1.0500000000000000e+00  0.0000000000000000e+00  0.0000000000000000e+00
1.0500000000000000e+00  5.5555555555555552e-02  0.0000000000000000e+00
1.0500000000000000e+00  9.5929535914450782e-03  0.0000000000000000e+00
1.0500000000000000e+00  2.7777777777777776e-02  0.0000000000000000e+00
1.0500000000000000e+00  4.5962601964110471e-02  0.0000000000000000e+00
9.4999999999999996e-01  1.1111111111111110e-01  0.0000000000000000e+00
9.4999999999999996e-01  6.5148509147000627e-02  0.0000000000000000e+00
```

so that I can directly identify the left and the right flap respectively.

**Comment by carme-hp** [comment]
Thanks! It is indeed smt related to paraview, when I open the files I see same as you:)

**Comment by MakisH** [comment]
@carme-hp can it maybe be that the files were out-of-date? What if you completely remove the `precice-output` directory and run again?

**Comment by MakisH** [comment]
> Should we unify the definition of `PF`, `PFleft` and `PFright` to avoid the code duplication? In the end only the location in x direction is different, right?

Good point. Maybe we could just make the location a parameter in the deal.II adapter?

**Comment by carme-hp** [comment]
@MakisH you are right! I could visualize it correctly when I completely removed the precice-output and run again. Sorry about the confusion. Maybe we could add the precice-output directory to the Allclean to avoid this problem happening again?

**Comment by MakisH** [comment]
> Maybe we could add the precice-output directory to the Allclean to avoid this problem happening again?

Yes, let's do that!

File: linear_elasticity/include/parameter_handling.h
Comment (MakisH) [inline linear_elasticity/include/parameter_handling.h:272]: What does the `(-3, 3)` mean here?

(I also assume that `"0.0"` is the default value.

**Review by MakisH — CHANGES_REQUESTED** [review]
Much cleaner and simpler than the previous attempt! :+1: :smile: 

Please remove the compiled binaries from the commit, add them to the `.gitignore`, and let's squash before merging.

Other than that, the code looks clear, apart from one point that may need a comment. @DavidSCN could you please also have a quick look at the code changes?

The linear case builds and seems to be running correctly and producing reasonable results visually. Same for the nonlinear.

Nice as well: this change is backwards-compatible. I did not need to update the tutorials to run it.

File: linear_elasticity/include/parameter_handling.h
Comment (DavidSCN) [reply linear_elasticity/include/parameter_handling.h:272]: These are upper and lower bounds for the specified values in the parameter file so that user can not exceed these positions.

File: linear_elasticity/include/parameter_handling.h
Comment (DavidSCN) [inline linear_elasticity/include/parameter_handling.h:289]: Maybe we can add "x" in the naming, so that it is clear for the user what's actually meant by this value.

File: nonlinear_elasticity/nonlinear_elasticity.cc
Comment (DavidSCN) [inline nonlinear_elasticity/nonlinear_elasticity.cc:481]: I think you replaced the wrong case. The "case order" of the if statement is not consistent with the linear case but you need to replace the other `if` branch of the PF (not FSI3).

**Review by DavidSCN — CHANGES_REQUESTED** [review]
I have (not yet) introduced a format checker, but I think the PR doesn't use the `.clang-format` of this repo. So please format the PR and maybe rebase your branch to the current develop branch.

**Comment by MakisH** [comment]
> I have (not yet) introduced a format checker, but I think the PR doesn't use the `.clang-format` of this repo. So please format the PR and maybe rebase your branch to the current develop branch.

@DavidSCN related issue: https://github.com/precice/dealii-adapter/issues/42

**Comment by DavidSCN** [comment]
> Nice as well: this change is backwards-compatible. I did not need to update the tutorials to run it.

Are you sure with this? If we change the parameter file, we also need to change it in the tutorials. Maybe we can define here a default value in order to avoid it. In addition, we need to update the parameter files in this repository.

**Comment by MakisH** [comment]
> > Nice as well: this change is backwards-compatible. I did not need to update the tutorials to run it.
> 
> Are you sure with this? If we change the parameter file, we also need to change it in the tutorials. Maybe we can define here a default value in order to avoid it. In addition, we need to update the parameter files in this repository.

I thought there was already a default value. If not, there should be. Does it not run out-of-the-box for you?

**Comment by carme-hp** [comment]
so I put 0.0 as the default value for the flat location, which is the location that we had before. So I expect it works backwards (it did for the flapperpendicular2D case)

**Comment by DavidSCN** [comment]
Ah ok, perfect!

**Review by MakisH — APPROVED** [review]
From my side, this can be merged after the files are formatted with clang-format.

@DavidSCN any last objections/wishes?

**Review by DavidSCN — APPROVED** [review]
No, not really. It might be worth to add the flap location to our 'default' parameter files in this repo in order to have a complete list of all available options here. However, you can do this according to your opinion. I can also format the file after merging if you struggle with this. Feel free to merge.

**Comment by DavidSCN** [comment]
...so that we can see the result of the indentation check..

**Pull request merged**
